pallets/usage/src/lib.rs

```diff
@@ -0,0 +1,384 @@
+        ///
+        /// Makes necessary actions to maintain data needed for charging programs
+        /// for keeping messages in the wait list
+        fn offchain_worker(now: BlockNumberFor<T>) {
+            // Only do something if we are a potential validator.
+            if !sp_io::offchain::is_validator() {
+                log::debug!(
+                    target: "gear-support",
+                    "Skipping offchain worker at {:?}: not a validator.",
+                    now,
+                );
+                return;
+            }
+
+            // Ensure we maintain minimum interval between full wait list traversals
+            let current_round_storage_ref =
+                StorageValueRef::persistent(offchain::STORAGE_ROUND_STARTED_AT);
+            let current_round_started_at =
+                match current_round_storage_ref.get::<BlockNumberFor<T>>() {
+                    Ok(maybe_round_started_at) => maybe_round_started_at.unwrap_or_default(),
+                    _ => {
+                        log::debug!(
+                            target: "gear-support",
+                            "Failed to get a value from storage at block {:?}",
+                            now,
+                        );
+                        return;
+                    }
+                };
+            let (_, last_key) = match offchain::get_last_key_from_offchain_storage() {
+                Ok(x) => x,
+                _ => {
+                    log::debug!(
+                        target: "gear-support",
+                        "Failed to get a value from storage at block {:?}",
+                        now,
+                    );
+                    return;
+                }
+            };
+
+            if now.saturating_sub(current_round_started_at) < T::WaitListTraversalInterval::get()
+                && &last_key[..] == common::STORAGE_WAITLIST_PREFIX
+            {
+                // We have either finished the previous round or never started one, and the number of
+                // elapsed blocks since last traversal is less than the expected minimum interval
+                log::debug!(
+                    target: "gear-support",
+                    "Block {:?} offchain worker. Not starting next wait list traversal until block {:?}",
+                    now,
+                    current_round_started_at.saturating_add(T::WaitListTraversalInterval::get()),
+                );
+                return;
+            }
+
+            if &last_key[..] == common::STORAGE_WAITLIST_PREFIX {
+                // Starting a new round
+                current_round_storage_ref.set(&now);
+            }
+
+            // Acquire the lock protecting shared offchain workers' storage
+            let mut lock = StorageLock::<'_, Time>::with_deadline(
+                offchain::STORAGE_OCW_LOCK,
+                Duration::from_millis(T::ExpirationDuration::get()),
+            );
+            let _guard = lock.lock();
```

Comment:
Can it finish with an error?

Reply:
It's a blocking call (if I understand correctly what you mean). There is another type of lock (`BlockAndTime`, to be specific) which returns an error in case of failure to acquire the lock. In case of the `Time` lock the next offchain worker will be able to take hold of it either after the previous holder drops it or after the lock expiry.
Using `Time` locks is more common, however, using the `BlockAndTime` lock is also an option, in case we decide it is worthwhile to do so.